pkg/controller/istiocsr/deployments.go

```diff
@@ -0,0 +1,382 @@
+	}
+	exist, err := r.Exists(r.ctx, key, fetched)
+	if err != nil {
+		return FromClientError(err, "failed to check %s deployment resource already exists", deploymentName)
+	}
+
+	if exist && istioCSRCreateRecon {
+		r.eventRecorder.Eventf(istiocsr, corev1.EventTypeWarning, "ResourceAlreadyExists", "%s deployment resource already exists, maybe from previous installation", deploymentName)
+	}
+	if exist && hasObjectChanged(desired, fetched) {
+		r.log.V(1).Info("deployment has been modified, updating to desired state", "name", deploymentName)
+		if err := r.UpdateWithRetry(r.ctx, desired); err != nil {
+			return FromClientError(err, "failed to update %s deployment resource", deploymentName)
+		}
+		r.eventRecorder.Eventf(istiocsr, corev1.EventTypeNormal, "Reconciled", "deployment resource %s reconciled back to desired state", deploymentName)
+	} else {
+		r.log.V(1).Info("deployment resource already exists and is in expected state", "name", deploymentName)
+	}
+	if !exist {
+		if err := r.Create(r.ctx, desired); err != nil {
+			return FromClientError(err, "failed to create %s deployment resource", deploymentName)
+		}
+		r.eventRecorder.Eventf(istiocsr, corev1.EventTypeNormal, "Reconciled", "deployment resource %s created", deploymentName)
+	}
+
+	if err := r.updateImageInStatus(istiocsr, desired); err != nil {
+		return FromClientError(err, "failed to update %s/%s istiocsr status with image info", istiocsr.GetNamespace(), istiocsr.GetName())
+	}
+	return nil
+}
+
+func (r *Reconciler) getDeploymentObject(istiocsr *v1alpha1.IstioCSR, resourceLabels map[string]string) (*appsv1.Deployment, error) {
+	if istiocsr.Spec.IstioCSRConfig == nil {
+		return nil, NewIrrecoverableError(fmt.Errorf("%s/%s spec.IstioCSRConfig is empty", istiocsr.GetNamespace(), istiocsr.GetName()), "not creating deployment resource")
+	}
+
+	if err := r.assertIssuerRefExists(istiocsr); err != nil {
+		return nil, fmt.Errorf("failed to verify issuer in %s/%s: %w", istiocsr.GetNamespace(), istiocsr.GetName(), err)
+	}
+
+	deployment := decodeDeploymentObjBytes(assets.MustAsset(deploymentAssetName))
+
+	updateNamespace(deployment, istiocsr.GetNamespace())
+	updateResourceLabels(deployment, resourceLabels)
+	updatePodTemplateLabels(deployment, resourceLabels)
+	updateResourceRequirement(deployment, istiocsr)
+	updateAffinityRules(deployment, istiocsr)
+	updatePodTolerations(deployment, istiocsr)
+	updateNodeSelector(deployment, istiocsr)
+	updateArgList(deployment, istiocsr)
+
+	if err := r.updateImage(deployment, istiocsr); err != nil {
+		return nil, NewIrrecoverableError(err, "failed to update image %s/%s", istiocsr.GetNamespace(), istiocsr.GetName())
+	}
+	if err := r.updateVolumes(deployment, istiocsr); err != nil {
+		return nil, fmt.Errorf("failed to update volume %s/%s: %w", istiocsr.GetNamespace(), istiocsr.GetName(), err)
+	}
+
+	return deployment, nil
+}
+
+func (r *Reconciler) updateImage(deployment *appsv1.Deployment, istiocsr *v1alpha1.IstioCSR) error {
```

Comment:
`istiocsr` object is not used in this function. Is there a reason it is added to the arg list?

Reply:
The earlier implementation made use of it, where status was updated in same method, but now its moved to different place.

Reply:
If it's not used then it can probably be removed from the arg list.

Reply:
suggestion incorporated.